pkg/operator/target_status.go

```diff
@@ -0,0 +1,936 @@
+			builder.WithPredicates(predicate.NewPredicateFuncs(func(o client.Object) bool {
+				return false
+			})),
+		).
+		Watches(&source.Channel{
+			Source: reconciler.ch,
+		}, &handler.EnqueueRequestForObject{}).
+		Complete(reconciler)
+	if err != nil {
+		return nil, errors.Wrap(err, "create target status controller")
+	}
+
+	return reconciler, nil
+}
+
+func newTargetStatusPoller(logger logr.Logger, opts Options, registry prometheus.Registerer, builder dataManagerBuilder) (*targetStatusReconciler, error) {
+	if err := registry.Register(targetStatusDuration); err != nil {
+		return nil, err
+	}
+
+	ch := make(chan event.GenericEvent, 1)
+
+	dataManager, err := builder.build()
+	if err != nil {
+		return nil, err
+	}
+
+	reconciler := &targetStatusReconciler{
+		ch:      ch,
+		opts:    opts,
+		manager: dataManager,
+		logger:  logger,
+	}
+
+	return reconciler, nil
+}
+
+type emptyObject struct{}
+
+func (r *targetStatusReconciler) Start(ctx context.Context, tickerBuilder func() Ticker) error {
+	r.tickerBuilder = tickerBuilder
+	r.queue()
+	return nil
+}
+
+// Queues a reconcile.
+func (r *targetStatusReconciler) queue() {
+	r.ch <- event.GenericEvent{
+		Object: &appsv1.DaemonSet{},
+	}
+}
+
+// Reconcile polls the collector pods, fetches and aggregates target status and upserts
+// into each PodMonitoring's Status field.
+func (r *targetStatusReconciler) Reconcile(ctx context.Context, _ reconcile.Request) (reconcile.Result, error) {
+	// minDelay := time.NewTimer(10 * time.Second)
+
+	minDelay := r.tickerBuilder()
+	// defer minDelay.Stop()
+
+	onceLog := new(sync.Once)
+
+	// Wrapper function so we could exit early.
+	err := func() error {
+		now := time.Now()
+		if err := poll(ctx, r.logger, r.opts, r.manager); err != nil {
+			// Not having data is not an error.
+			if apierrors.IsNotFound(err) {
+				onceLog.Do(func() {
+					r.logger.Error(err, "Kubernetes config not found")
+				})
+				return nil
+			}
+			return err
+		}
+		// In case Kubernetes config is removed and re-added.
+		onceLog = new(sync.Once)
+
+		duration := time.Since(now)
+		targetStatusDuration.WithLabelValues().Set(float64(duration.Milliseconds()))
+		return nil
+	}()
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Check if we beat the timer, otherwise wait.
+	<-minDelay.Channel()
+	// Re-trigger event
+	r.queue()
+
+	return reconcile.Result{}, nil
+}
+
+func poll(ctx context.Context, logger logr.Logger, opts Options, manager dataManager) error {
+	targets, err := fetchTargets(ctx, logger, opts, manager)
+	if err != nil {
+		return err
+	}
+
+	return populateTargets(ctx, logger, manager, targets)
+}
+
+func fetchTargets(ctx context.Context, logger logr.Logger, opts Options, manager dataManager) ([]*prometheusTargetsResultAndTime, error) {
+	podMap, err := manager.getPrometheusPods(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Set up pod job queue and jobs
+	podDiscoveryChan := make(chan prometheusPod, opts.TargetPollConcurrency)
```

Comment:
Did you see any performance gain when using the buffer?

I feel like a more conventional use of buffers is to allow the sender to batch a bunch of requests and then do other stuff. However, in this case the sender is only concerned with putting items in the channel.

Your concurrency is controlled in the for loop, with a fixed set of blocking goroutines, each of which are continually trying to pull new items from the pod channel. Since the sender has virtually no latency in sending more items through the channel (i.e. the receivers aren't going to "beat" the sender before it can send in a new signal) an unbounded channel should be fine.